File: server/inference/gps_sync.py
"""Map video frame time (ms since epoch) to latitude/longitude from gps_log.json."""

from __future__ import annotations

from typing import Any


def lat_lon_at_timestamp(ts_ms: float, points: list[dict[str, Any]]) -> tuple[float, float] | None:
    """
    Linear interpolation between GPS samples. Timestamps are epoch milliseconds.
    Returns (latitude, longitude) or None if no points.
    """
    if not points:
        return None

    pts = sorted(points, key=lambda p: float(p["timestamp"]))

    if ts_ms <= float(pts[0]["timestamp"]):
        return float(pts[0]["latitude"]), float(pts[0]["longitude"])
    if ts_ms >= float(pts[-1]["timestamp"]):
        return float(pts[-1]["latitude"]), float(pts[-1]["longitude"])

    for i in range(len(pts) - 1):
        t0 = float(pts[i]["timestamp"])
        t1 = float(pts[i + 1]["timestamp"])
        if t0 <= ts_ms <= t1:
            if t1 == t0:
                a = 0.0
            else:
                a = (ts_ms - t0) / (t1 - t0)
            lat = float(pts[i]["latitude"]) + a * (float(pts[i + 1]["latitude"]) - float(pts[i]["latitude"]))
            lon = float(pts[i]["longitude"]) + a * (float(pts[i + 1]["longitude"]) - float(pts[i]["longitude"]))
            return lat, lon

    return float(pts[-1]["latitude"]), float(pts[-1]["longitude"])


def format_pothole_id(longitude: float, latitude: float) -> str:
    """
    Filename-safe id: potholeid_<long>_<lat> with . → p and - → m
    (longitude first, then latitude).
    """
    def enc(v: float) -> str:
        return f"{v:.6f}".replace("-", "m").replace(".", "p")

    return f"potholeid_{enc(longitude)}_{enc(latitude)}"

File: services/gpsSync.js
/**
 * Port of server/inference/gps_sync.py for JS.
 * @param {number} tsMs
 * @param {Array<{ timestamp: number, latitude: number, longitude: number }>} points
 * @returns {{ latitude: number, longitude: number } | null}
 */
export function latLonAtTimestamp(tsMs, points) {
  if (!points?.length) {
    return null;
  }

  const pts = [...points].sort((a, b) => Number(a.timestamp) - Number(b.timestamp));
  const t = Number(tsMs);

  if (t <= Number(pts[0].timestamp)) {
    return { latitude: Number(pts[0].latitude), longitude: Number(pts[0].longitude) };
  }
  if (t >= Number(pts[pts.length - 1].timestamp)) {
    const last = pts[pts.length - 1];
    return { latitude: Number(last.latitude), longitude: Number(last.longitude) };
  }

  for (let i = 0; i < pts.length - 1; i++) {
    const t0 = Number(pts[i].timestamp);
    const t1 = Number(pts[i + 1].timestamp);
    if (t0 <= t && t <= t1) {
      const denom = t1 - t0;
      const a = denom === 0 ? 0 : (t - t0) / denom;
      const lat =
        Number(pts[i].latitude) + a * (Number(pts[i + 1].latitude) - Number(pts[i].latitude));
      const lon =
        Number(pts[i].longitude) + a * (Number(pts[i + 1].longitude) - Number(pts[i].longitude));
      return { latitude: lat, longitude: lon };
    }
  }

  const last = pts[pts.length - 1];
  return { latitude: Number(last.latitude), longitude: Number(last.longitude) };
}

/**
 * @param {number} longitude
 * @param {number} latitude
 */
export function formatPotholeId(longitude, latitude) {
  const enc = (v) =>
    v.toFixed(6).replace(/-/g, 'm').replace(/\./g, 'p');
  return `potholeid_${enc(longitude)}_${enc(latitude)}`;
}
